### PR TITLE
whitelist signature scheme in config

### PR DESCRIPTION
**Pull request by dustinxie**
private/public key operation will verify against the whitelist schemes in config before continuing

### CODE REVIEW

File: blockchain/block/header.go [hunk 1]
```diff
@@ -166,7 +165,7 @@ func (h *Header) HashHeaderCore() hash.Hash256 {
 func (h *Header) VerifySignature() bool {
 	hash := h.HashHeaderCore()
 
-	if h.pubkey == nil || len(h.blockSig) != action.SignatureLength {
+	if h.pubkey == nil {
```
Comment (dustinxie) [inline blockchain/block/header.go]: this check will be done inside pubkey.Verify() below

File: ioctl/cmd/account/account.go [hunk 2]
```diff
@@ -235,7 +234,7 @@ func storeKey(privateKey, walletDir, password string) (string, error) {
 			return "", output.NewError(output.KeystoreError, "failed to import private key into keystore ", err)
 		}
 	case *crypto.P256sm2PrvKey:
-		if _, err := sm2.WritePrivateKeytoPem(walletDir, sk.PrivateKey, []byte(password)); err != nil {
+		if err := crypto.WritePrivateKeyToPem(walletDir, sk, password); err != nil {
```
Comment (dustinxie) [inline ioctl/cmd/account/account.go]: using crypto func to avoid directly importing sm2

File: config/config.go [hunk 3]
```diff
@@ -499,9 +509,50 @@ func (cfg Config) ProducerPrivateKey() crypto.PrivateKey {
...
+	for _, e := range cfg.Chain.SignatureScheme {
+		if sigScheme == e {
+			// signature scheme is whitelisted
+			return true
+		}
+	}
+	return false
+}
+
+func (cfg Config) generateRandomKey() string {
```
Comment (CoderZhi) [inline config/config.go]: this function confused me. How could we randomly pick a scheme to generate a random key?

File: config/config.go [hunk 4]
```diff
@@ -224,6 +229,7 @@ type (
 		ID              uint32           `yaml:"id"`
 		Address         string           `yaml:"address"`
 		ProducerPrivKey string           `yaml:"producerPrivKey"`
+		SignatureScheme []string         `yaml:"signatureScheme"`
```
Comment (CoderZhi) [inline config/config.go]: I thought we start with one valid scheme

Comment (dustinxie) [reply config/config.go]: if the chain wants only 1 valid scheme, then put only 1 in the config

the default is []string{SigP256k1}, which is the current scheme supported

File: config/config.go (hunk 3 above)
Comment (dustinxie) [reply config/config.go]: this is to cover the corner case cfg.Chain.ProducerPrivKey is left empty, then generate a random key according to allowed schemes.

i think i'll leave it as empty, and not change the current code's behavior, please see the latest commit

File: config/config.go [hunk 5]
```diff
@@ -423,9 +420,6 @@ func New(validates ...Validate) (Config, error) {
 		return Config{}, errors.Wrap(err, "failed to unmarshal YAML config to struct")
 	}
 
-	if cfg.Chain.ProducerPrivKey == "" {
```
Comment (dustinxie) [inline config/config.go]: remove it so not change original code's behavior

**Pull request merged**